### PR TITLE
RSDK-4771: virtual base station connection

### PR DESCRIPTION
**Pull request by susmitaSanyal**
This PR includes some additional function to get NTRIP stream from virtual reference station. We will always get the source table and parse it based on the mount point in config, but depending on the nmea bit from the source table we will either send GGA messages to caster or not.

Will move common functions to rdkutil in another PR.

### CODE REVIEW

**Review by randhid — CHANGES_REQUESTED** [review]

File: components/movementsensor/gpsrtkserial/gpsrtkserial.go [hunk 1]
```diff
@@ -637,3 +657,77 @@ func (g *rtkSerial) Close(ctx context.Context) error {
...
+// from a virtual reference point.
+func (g *rtkSerial) sendGGAMessage() {
+	// Open the serial port for reading
+	serialPort, err := slib.Open(slib.OpenOptions{
+		PortName:        g.writePath,
+		BaudRate:        uint(g.wbaud),
+		DataBits:        8,
+		StopBits:        1,
+		MinimumReadSize: 1,
+	})
```
Comment (randhid) [inline components/movementsensor/gpsrtkserial/gpsrtkserial.go:671]: If these are commonly shared with any other part of the code in this package, let's set them as constants at the top of the file. It may be worth it to consider this for an eventual refactor, since I believe some of these rates and serialibrayr options rely on the underlying `nmeaMovementSensor` implementation, which we shouldn't expose in this driver long term.

My vision for this is just to have the nmea data parser be the package with ANY exposed export functions. And we'd rely on each GPS driver to read an write to a serial port. But too much work for now.

File: components/movementsensor/gpsnmea/nmeaParser.go [hunk 2]
```diff
@@ -238,8 +238,9 @@ func (g *GPSData) updateGNS(gns nmea.GNS) error {
 	return nil
 }
 
-//nolint:all
 // updateHDT updaates g.CompassHeading with the ground speed information from the provided
+//
+//nolint:all
```
Comment (penguinland) [inline components/movementsensor/gpsnmea/nmeaParser.go:243]: Eh... any chance this can be made more specific? Which lint rules are we trying to skip here?

I think of `nolint:all` as being a quick-and-dirty thing while you're debugging, and the ones that end up in the final code should be only the rules we're skipping on purpose.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 3]
```diff
@@ -354,6 +355,15 @@ func (g *rtkI2C) receiveAndWriteI2C(ctx context.Context) {
 		g.logger.Infof("caster %s seems to be down", g.ntripClient.URL)
 	}
 
+	srctable, err := g.ntripClient.Client.ParseSourcetable()
```
[the remark this answers is not in the document]
Comment (penguinland) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:358]: ...and while we're at it, perhaps `ParseSourceTable` as well.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 4]
```diff
@@ -354,6 +355,15 @@ func (g *rtkI2C) receiveAndWriteI2C(ctx context.Context) {
...
+	if err != nil {
+		g.logger.Errorf("failed to get source table: %v", err)
```
Comment (penguinland) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:360]: Naive question from someone who doesn't know Go well: shouldn't you use `%w` to wrap an error? When do you use that, and when `%v`?

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 5]
```diff
@@ -354,6 +355,15 @@ func (g *rtkI2C) receiveAndWriteI2C(ctx context.Context) {
...
+	}
+	nmea, nmeaerr := findLineWithMountPoint(srctable, g.ntripClient.MountPoint)
+	if nmeaerr != nil {
+		g.logger.Errorf("can't find mountpoint in source table, found err %v\n", nmeaerr)
```
Comment (penguinland) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:364]: Same question here, but also I wouldn't put a newline in the error text, because that will make the logging look bad.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 6]
```diff
@@ -354,6 +355,15 @@ func (g *rtkI2C) receiveAndWriteI2C(ctx context.Context) {
 		g.logger.Infof("caster %s seems to be down", g.ntripClient.URL)
 	}
 
+	srctable, err := g.ntripClient.Client.ParseSourcetable()
+	if err != nil {
+		g.logger.Errorf("failed to get source table: %v", err)
+	}
+	nmea, nmeaerr := findLineWithMountPoint(srctable, g.ntripClient.MountPoint)
```
Comment (penguinland) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:362]: Same suggestion as Rand had earlier: `nmeaErr` is way easier to read.

File: components/movementsensor/gpsrtkserial/gpsrtkserial.go [hunk 7]
```diff
@@ -417,6 +432,10 @@ func (g *rtkSerial) receiveAndWriteSerial() {
 			return
 		default:
 		}
+		if nmea {
+			// only send GGA messages if nmea bit is 1 in source table.
```
[the remark this answers is not in the document]
Comment (penguinland) [reply components/movementsensor/gpsrtkserial/gpsrtkserial.go:436]: Rand, thanks for saying that. I had no idea what this meant, but I had assumed it was obvious to people who know GPS stuff. Background context is important!

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 8]
```diff
@@ -682,3 +698,80 @@ func (g *rtkI2C) Close(ctx context.Context) error {
...
+func (g *rtkI2C) sendGGAMessage() {
+	// Open an I2C handle to read NMEA messages
+	handle, err := g.bus.OpenHandle(g.addr)
+	if err != nil {
+		g.logger.Errorf("can't open gps i2c handle: %s", err)
+		g.err.Set(err)
+		return
+	}
+	defer func() {
+		if closeErr := handle.Close(); closeErr != nil {
```
Comment (penguinland) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:713]: :+1: Thanks for deferring closing the handle immediately! This avoids all kinds of bugs that might otherwise occur.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 9]
```diff
@@ -682,3 +698,80 @@ func (g *rtkI2C) Close(ctx context.Context) error {
...
+		// Read from the I2C device
+		buffer, err := handle.Read(g.cancelCtx, 1024)
+		if err != nil {
+			g.err.Set(err)
+			return
+		}
+
+		// Process the received data
+		for _, b := range buffer {
+			// PMTK uses CRLF line endings to terminate sentences, but just LF to blank data.
```
Comment (penguinland) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:738]: > just LF to blank data.

Did you mean "just LF _for_ blank data"? or "just LF to _clear_ data"? I can't tell what this means. The next line makes more sense: I wonder if it should be

> PMTK uses CRLF line endings at the end of sentences, but LF to distinguish data in the middle. We ignore line feeds and look for the carriage return to mark the end of the sentence.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 10]
```diff
@@ -682,3 +698,80 @@ func (g *rtkI2C) Close(ctx context.Context) error {
...
+			// Since CR should never appear except at the end of our sentence, we use that to determine sentence end.
+			// LF is merely ignored.
+			if b == 0x0D {
```
Comment (penguinland) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:741]: Put in a comment explaining that 0x0D is ASCII for carriage return. ...and maybe repeat here that the carriage return means we've reached the end of a sentence.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 11]
```diff
@@ -682,3 +698,80 @@ func (g *rtkI2C) Close(ctx context.Context) error {
...
+					// Convert the received bytes to a string
+					receivedData := string(messageBuffer)
+
+					// Check for NMEA GGA messages
+					messages := strings.Split(receivedData, "\n")
+					for _, message := range messages {
+						if strings.Contains(message, "GGA") {
+							// Send the GGA message to the NTRIP caster
+							if _, err := g.correctionWriter.Write([]byte(message)); err != nil {
+								g.logger.Errorf("error while sending GGA message to NTRIP caster: %v", err)
```
Comment (penguinland) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:752]: This is nested 8 layers deep. I wonder if there's a way to make this code flatter, possibly by breaking out a smaller piece into a helper function.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 12]
```diff
@@ -682,3 +698,80 @@ func (g *rtkI2C) Close(ctx context.Context) error {
...
+								g.err.Set(err)
+								return
+							}
+						}
+					}
+				}
+				messageBuffer = messageBuffer[:0]
+			} else if b != 0x0A && b != 0xFF { // adds only valid bytes
```
Comment (penguinland) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:760]: I'd rephrase the comment, to note that 0x0A is a newline, and we ignore all of these. 

Why are we skipping 0xFF? What does that represent?

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 13]
```diff
@@ -682,3 +698,80 @@ func (g *rtkI2C) Close(ctx context.Context) error {
...
+			// PMTK uses CRLF line endings to terminate sentences, but just LF to blank data.
+			// Since CR should never appear except at the end of our sentence, we use that to determine sentence end.
+			// LF is merely ignored.
+			if b == 0x0D {
+				if len(messageBuffer) > 0 {
+					// Convert the received bytes to a string
+					receivedData := string(messageBuffer)
+
+					// Check for NMEA GGA messages
+					messages := strings.Split(receivedData, "\n")
```
Comment (penguinland) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:747]: I thought line 760 was already filtering out newlines. I wonder if we can get rid of that and keep this one instead, because this one is easier to understand.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 14]
```diff
@@ -682,3 +698,80 @@ func (g *rtkI2C) Close(ctx context.Context) error {
...
+								g.logger.Errorf("error while sending GGA message to NTRIP caster: %v", err)
+								g.err.Set(err)
+								return
+							}
+						}
+					}
+				}
+				messageBuffer = messageBuffer[:0]
+			} else if b != 0x0A && b != 0xFF { // adds only valid bytes
+				messageBuffer = append(messageBuffer, b)
```
Comment (penguinland) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:761]: You're processing this character-by-character, but I suspect it would be simpler to just split on carriage returns.

```go
// messageBuffer is left over from last time
buffer, err := messageBuffer + handle.Read(g.cancelCtx, 1024)
// Strip out newlines and whatever 0xFF is
buffer = strings.replace(strings.replace(buffer, "\n", ""), "\xFF", "")
pieces := strings.Split(buffer, "\r")
// The last piece didn't end with a carriage return yet. Keep it around because it's not finished.
messageBuffer = pieces[len(pieces)-1] // To be prepended to the buffer next time
// All other pieces are complete messages
for _, message := range pieces[:len(pieces)-1] {
    if strings.Contains(message, "GGA") {
        ...
    }
}
```

I think that should remove several levels of indentation, and make the algorithm easier to understand because we're using higher-level functions like split and replace instead of going character-by-character.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 15]
```diff
@@ -682,3 +698,80 @@ func (g *rtkI2C) Close(ctx context.Context) error {
...
+							// Send the GGA message to the NTRIP caster
+							if _, err := g.correctionWriter.Write([]byte(message)); err != nil {
+								g.logger.Errorf("error while sending GGA message to NTRIP caster: %v", err)
+								g.err.Set(err)
+								return
+							}
+						}
+					}
+				}
+				messageBuffer = messageBuffer[:0]
```
Comment (penguinland) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:759]: The other file has a comment explaining that this is how to clear the buffer. Put the comment here too; I wasn't sure if there was a typo on this line.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go (hunk 4 above)
Comment (randhid) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:360]: `%v` is the styling for anything when you're trying to log, rather than `%w` which is specifically for an error. 

I do like `%v` since it takes the load off the variable changing in the future, or thinking a variable is one type and then figuring out through your logs after it is in production that it is in fact another.  But your mileage may vary on using the appropriate wrapper when you know for sure the wrapper is a specific type, as in this case where we know for sure it is an error.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 16]
```diff
@@ -682,3 +698,80 @@ func (g *rtkI2C) Close(ctx context.Context) error {
 
 	return nil
 }
+
+// sendGGAMessage sends GGA messages to the serial port. This is only used to get NTRIP stream
+// from a virtual reference point.
+func (g *rtkI2C) sendGGAMessage() {
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:704]: Yes, this can't be shared, but some other functions can be, I'll create a ticket and link it here.

File: components/movementsensor/gpsrtkserial/gpsrtkserial.go [hunk 17]
```diff
@@ -637,3 +657,77 @@ func (g *rtkSerial) Close(ctx context.Context) error {
...
+					return
+				}
+			}
+		}
+
+		// Clear the message buffer
+		messageBuffer = messageBuffer[:0]
+
+		// Sleep for a while before reading again
+		time.Sleep(time.Second) // Adjust the sleep duration as needed
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkserial/gpsrtkserial.go:719]: I do need to test it out a little more to find the optimal wait time. left the comment so I don't forget about this. I do believe sending GGA messages periodically is necessary in this case.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:358]: @penguinland `ParseSourceTable` come from outside of RDK. It's part of the ntrip client package that we use.

[the remark this answers is not in the document]
Comment (penguinland) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:358]: (Rand's comment got marked as resolved, but I don't see the change as committed yet. Not sure if it's on your local branch but not pushed to Github yet, or if something else went wrong.)

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go [hunk 18]
```diff
@@ -682,3 +698,80 @@ func (g *rtkI2C) Close(ctx context.Context) error {
...
+		select {
+		case <-g.cancelCtx.Done():
+			return
+		default:
+		}
+
+		// Read from the I2C device
+		buffer, err := handle.Read(g.cancelCtx, 1024)
+		if err != nil {
+			g.err.Set(err)
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk.go:732]: We don't need to explicitly call close. Deferred call will take care of it.

**Review by randhid — APPROVED** [review]
Let's hardware test out the new changes once last time, make sure i2c works and get it in.

File: components/movementsensor/gpsrtkserial/gpsrtkserial.go [hunk 19]
```diff
@@ -637,3 +659,77 @@ func (g *rtkSerial) Close(ctx context.Context) error {
...
+		messageBuffer = messageBuffer[:0]
+
+		// Sleep for a while before reading again.
+		time.Sleep(time.Second)
+	}
+}
+
+// findLineWithMountPoint parses the given source-table returns the nmea bool of the given mount point.
+// TODO: RSDK-5462.
+func findLineWithMountPoint(sourceTable *ntrip.Sourcetable, mountPoint string) (bool, error) {
```
Comment (randhid) [inline components/movementsensor/gpsrtkserial/gpsrtkserial.go:727]: Consider moving this to rtkutils

**Comment by susmitaSanyal** [comment]
Debugging this some more. Going to close this PR for now. will reopen when done debugging.
